third/cns/src/man/ksu.1
.\" Copyright (c) 1988 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms are permitted
.\" provided that the above copyright notice and this paragraph are
.\" duplicated in all such forms and that any documentation,
.\" advertising materials, and other materials related to such
.\" distribution and use acknowledge that the software was developed
.\" by the University of California, Berkeley.  The name of the
.\" University may not be used to endorse or promote products derived
.\" from this software without specific prior written permission.
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
.\" WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\"	@(#)su.1	6.7 (Berkeley) 12/7/88
.\"
.TH KSU 1 "Kerberos Version 4.0" "MIT Project Athena"
.UC
.SH NAME
ksu \- substitute user id, using Kerberos
.SH SYNOPSIS
.B ksu
[-flm] [login]
.SH DESCRIPTION
\fIKsu\fP requests the password for \fIlogin\fP (or for ``root'', if no
login is provided), and switches to that user and group ID.  A shell is
then invoked.
.PP
By default, your environment is unmodified with the exception of
\fIUSER\fP, \fIHOME\fP, and \fISHELL\fP.  \fIHOME\fP and \fISHELL\fP
are set to the target login's \fI/etc/passwd\fP values.  \fIUSER\fP
is set to the target login, unless the target login has a UID of 0,
in which  case it is unmodified.  The invoked shell is the target
login's.  This is the traditional behavior of \fIksu\fP.
.PP
The \fI-l\fP option simulates a full login.  The environment is discarded
except for \fIHOME\fP, \fISHELL\fP, \fIPATH\fP, \fITERM\fP, and \fIUSER\fP.
\fIHOME\fP and \fISHELL\fP are modified as above.  \fIUSER\fP is set to
the target login.  \fIPATH\fP is set to ``/usr/ucb:/bin:/usr/bin''.
\fITERM\fP is imported from your current environment.  The invoked shell
is the target login's, and \fIksu\fP will change directory to the target
login's home directory.
.PP
The \fI-m\fP option causes the environment to remain unmodified, and
the invoked shell to be your login shell.  No directory changes are
made.  As a security precaution, if the 
.I -m
option is specified, the target user's shell is a non-standard shell
(as defined by \fIgetusershell\fP(3)) and the caller's real uid is
non-zero,
.I su
will fail.
.PP
If the invoked shell is \fIcsh\fP, the \fI-f\fP option prevents it from
reading the \fI.cshrc\fP file.  Otherwise, this option is ignored.
.PP
Only users with root instances listed in /\&.klogin may \fIksu\fP to
``root''  (The format of this file is described by \fIrlogin\fP(1).).  When
attempting root access, \fIksu\fP attempts to fetch a
ticket-granting-ticket for ``username.root@localrealm'', where
\fIusername\fP is the username of the process.  If possible, the tickets
are used to obtain, use, and verify tickets for the service
``rcmd.host@localrealm'' where \fIhost\fP is the canonical host name (as
determined by 
.IR krb_get_phost (3))
of the machine.  If this verification
fails, the \fIksu\fP is disallowed (If the service
``rcmd.host@localrealm'' is not registered, the \fIksu\fP is allowed.).
.PP
By default (unless the prompt is reset by a startup file) the super-user
prompt is set to ``#'' to remind one of its awesome power.
.PP
When not attempting to switch to the ``root'' user,
.I ksu
behaves exactly like
.IR su (1).
.SH "SEE ALSO"
su(1), csh(1), login(1), rlogin(1), sh(1), krb_get_phost(3), passwd(5),
group(5), environ(7)


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/cns/src/kuser/ksu.c
/*
 * ksu.c
 *
 * Copyright (c) 1988 The Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that the above copyright notice and this paragraph are
 * duplicated in all such forms and that any documentation,
 * advertising materials, and other materials related to such
 * distribution and use acknowledge that the software was developed
 * by the University of California, Berkeley.  The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 * Kerberos additions Copyright 1987, 1988 by the Massachusetts Institute
 * of Technology. For copying and distribution information, please see
 * the file <mit-copyright.h>.
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1988 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)su.c	5.11 (Berkeley) 12/7/88";
#endif /* not lint */

#include "mit-copyright.h"
#include <sys/types.h>
#include <sys/param.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <syslog.h>
#include <stdio.h>
#include <string.h>
#include <pwd.h>
#include <grp.h>
#include "krb.h"
#include <netdb.h>
#include <sys/ioctl.h>

#ifdef USE_UNISTD_H
#include <unistd.h>
#endif

#ifndef LOG_AUTH
#define LOG_AUTH 0
#endif /* LOG_AUTH */

/* for Ultrix and friends ... */
#ifndef MAXHOSTNAMELEN
#define MAXHOSTNAMELEN 64
#endif

int     kerno;
char    lrealm[REALM_SZ];
char    krbtkfile[128];

#define MAXPWSIZE	128	/* Biggest string we accept for a password
				   (includes space for null terminator) */

#ifdef __svr4__
#include <time.h>
#endif

void svr4_log (from, to, tty, ok)
	char *from, *to, *tty;
	int ok;
{
#ifdef __svr4__
	time_t now;
	struct tm *tm;
	FILE *f;

	setuid (0);		/* we're only called just before exiting, or on success */
	f = fopen ("/var/adm/sulog", "a");
	if (f == NULL) {
		perror ("/var/adm/sulog");
		return;
	}
	time (&now);
	tm = localtime (&now);
	if (!strncmp ("/dev/", tty, 5))
		tty += 5;
	if (tm)
		fprintf (f, "KSU %02d/%02d %02d:%02d %c %s %s-%s\n",
			 tm->tm_mon + 1, tm->tm_mday, tm->tm_hour, tm->tm_min,
			 ok ? '+' : '-',
			 tty, from, to);
	else
		perror ("localtime");
	fclose (f);
#endif
}

int main(argc, argv)
	int argc;
	char **argv;
{
	extern char **environ;
	extern int errno, optind;
	register struct passwd *pwd;
	register char *p, **g;
	struct group *gr;
	uid_t ruid, getuid();
#ifdef NO_GETUSERSHELL
	int ch, fulllogin, fastlogin, prio;
#else
	int asme, ch, fulllogin, fastlogin, prio;
#endif /* NO_GETUSERSHELL */
	int use_preauth = 0;
	enum { UNSET, YES, NO } iscsh = UNSET;
	char *user, *shell, *username, *cleanenv[2], *nargv[4], **np;
	char namebuf[50], shellbuf[MAXPATHLEN];
	char *crypt(), *getpass(), *getenv(), *getlogin();
#ifdef NOENCRYPTION
#define read_long_pw_string placebo_read_pw_string
#else
#define read_long_pw_string des_read_pw_string
#endif
	int read_long_pw_string();
	char pw_buf[MAXPWSIZE];
	char   *mytty;
  
	mytty = isatty(2) ? (char *) ttyname(2) : "(no tty)";

	np = &nargv[3];
	*np-- = NULL;
#ifdef NO_GETUSERSHELL
	fulllogin = fastlogin = 0;
#define	GETOPTARG "-flp"
#else
	asme = fulllogin = fastlogin = 0;
#define GETOPTARG "-flmp"
#endif
	while ((ch = getopt(argc, argv, GETOPTARG)) != EOF)
		switch((char)ch) {
		case 'f':
			fastlogin = 1;
			break;
		case 'p':
			use_preauth = 1;
			break;
		case '-':
		case 'l':
			fulllogin = 1;
			break;
#ifndef NO_GETUSERSHELL
		case 'm':
			asme = 1;
			break;
#endif
		case '?':
		default:
#ifdef NO_GETUSERSHELL
			fprintf(stderr, "usage: ksu [-flp] [login]\n");
#else
			fprintf(stderr, "usage: ksu [-flmp] [login]\n");
#endif /* NO_GETUSERSHELL */
			exit(1);
		}
	argv += optind;

	errno = 0;
#ifndef NO_SETPRIORITY
	prio = getpriority(PRIO_PROCESS, 0);
	if (errno)
		prio = 0;
	(void)setpriority(PRIO_PROCESS, 0, -2);
#endif

	/* get current login name and shell */
	if ((pwd = getpwuid(ruid = getuid())) == NULL) {
		fprintf(stderr, "ksu: who are you?\n");
		exit(1);
	}
	username = strcpy(namebuf, pwd->pw_name);
#ifndef NO_GETUSERSHELL
	if (asme)
		if (pwd->pw_shell && *pwd->pw_shell)
			shell = strcpy(shellbuf,  pwd->pw_shell);
		else {
			shell = "/bin/sh";
			iscsh = NO;
		}
#endif

	/* get target login information */
	user = *argv ? *argv : "root";
	if ((pwd = getpwnam(user)) == NULL) {
		fprintf(stderr, "ksu: unknown login %s\n", user);
		exit(1);
	}

	/*
	 * Only allow those with Kerberos "foo" instances in the ~foo/.klogin
	 * file to su to "foo".
	 */
	{
	    KTEXT_ST ticket;
	    AUTH_DAT authdata;
	    char hostname[MAXHOSTNAMELEN], savehost[MAXHOSTNAMELEN];
	    char promptbuf[100];
	    unsigned long faddr;
	    struct hostent *hp;

	    /* First lets see if he has a chance! */
	    if (krb_get_lrealm(lrealm, 1) != KSUCCESS) {
		fprintf(stderr,"Unable to get local realm\n");
		svr4_log (username, user, mytty, 0);
		exit(1);
	    }
	    if (koktologin(username, lrealm, user)) {
		fprintf(stderr,"You are not allowed to ksu to %s\n", user);
		svr4_log (username, user, mytty, 0);
		exit(1);
	    }
	    sprintf(krbtkfile, "/tmp/tkt_%s_%d", user, getuid());
	    setuid(0);		/* so ticket file has good protection */
	    sprintf (promptbuf, "Your `%s' instance password: ", user);
	    if (read_long_pw_string(pw_buf, sizeof(pw_buf)-1, promptbuf, 0)) {
		fprintf(stderr,"Error reading password.\n");
		svr4_log (username, user, mytty, 0);
		exit(1);
	    }
	    p = pw_buf;
	    setenv("KRBTKFILE", krbtkfile, 1);
	    if (use_preauth)
		kerno = krb_get_pw_in_tkt_preauth(username, user, lrealm,
						  "krbtgt", lrealm, 2, p);
	    else
		kerno = krb_get_pw_in_tkt(username, user, lrealm, "krbtgt",
				 	  lrealm, 2, p);
	    memset(p, 0, strlen(p));
	    if (kerno != KSUCCESS) {
		printf("Unable to ksu: %s\n", krb_get_err_text (kerno));
		syslog(LOG_NOTICE|LOG_AUTH, "ksu: BAD SU %s on %s: %s",
		       username, mytty, krb_get_err_text (kerno));
		svr4_log (username, user, mytty, 0);
		exit(1);
	    }
#ifndef NO_SETPRIORITY
	    setpriority(PRIO_PROCESS, 0, -2);
#endif
	    /*
	     * Now use the ticket for something useful, to make sure
	     * it is valid.
	     */
	    if (gethostname(hostname, sizeof(hostname)) == -1) {
		perror("cannot retrieve hostname");
		dest_tkt();
		svr4_log (username, user, mytty, 0);
		exit(1);
	    }
	    (void) strncpy(savehost, krb_get_phost(hostname),
			   sizeof(savehost));
	    savehost[sizeof(savehost)-1] = 0;

	    kerno = krb_mk_req(&ticket, "rcmd", savehost, lrealm, 33);
	    if (kerno == KDC_PR_UNKNOWN) {
		printf("Warning: tgt not verified\n");
		syslog(LOG_NOTICE|LOG_AUTH, "ksu: %s on %s: tgt not verified",
		       username, mytty);
	    } else if (kerno != KSUCCESS) {
		printf("Unable to use tgt: %s\n", krb_get_err_text (kerno));
		syslog(LOG_NOTICE|LOG_AUTH, "ksu: failed su: %s on %s: %s",
		       username, mytty, krb_get_err_text (kerno));
		dest_tkt();
		svr4_log (username, user, mytty, 0);
		exit(1);
	    } else {
		if (!(hp = gethostbyname(hostname))) {
		    printf("Unable to get address of %s\n",hostname);
		    dest_tkt();
		    svr4_log (username, user, mytty, 0);
		    exit(1);
		}
		memcpy((char *) &faddr, (char *)hp->h_addr, sizeof(faddr));
		if ((kerno = krb_rd_req(&ticket, "rcmd", savehost,
					faddr, &authdata, "")) != KSUCCESS) {
		    printf("Unable to verify rcmd ticket: %s\n",
			   krb_get_err_text (kerno));
		    syslog(LOG_NOTICE|LOG_AUTH, "ksu: failed su: %s on %s: %s",
			   username, mytty, krb_get_err_text (kerno));
		    dest_tkt();
		    svr4_log (username, user, mytty, 0);
		    exit(1);
		}		
	    }
	    printf("Don't forget to kdestroy before exiting the new shell.\n");
	    svr4_log (username, user, mytty, 1);
	}

#ifndef NO_GETUSERSHELL
	if (asme) {
		/* if asme and non-standard target shell, must be root */
		if (!chshell(pwd->pw_shell) && ruid) {
			fprintf(stderr, "ksu: Permission denied.\n");
			dest_tkt();
			exit(1);
		}
	}
	else
#endif
	if (pwd->pw_shell && *pwd->pw_shell) {
		shell = pwd->pw_shell;
		iscsh = UNSET;
	} else {
		shell = "/bin/sh";
		iscsh = NO;
	}

	/* if we're forking a csh, we want to slightly muck the args */
	if (iscsh == UNSET) {
		if (p = strrchr(shell, '/'))
			++p;
		else
			p = shell;
		iscsh = strcmp(p, "csh") ? NO : YES;
	}

	/* set permissions */
	if (setgid(pwd->pw_gid) < 0) {
		perror("ksu: setgid");
		dest_tkt();
		exit(1);
	}
#ifndef __SCO__
	if (initgroups(user, pwd->pw_gid)) {
		fprintf(stderr, "ksu: initgroups failed\n");
		dest_tkt();
		exit(1);
	}
#endif
	if (setuid(pwd->pw_uid) < 0) {
		perror("ksu: setuid");
		dest_tkt();
		exit(1);
	}

#ifndef NO_GETUSERSHELL
	if (!asme) {
#endif
		if (fulllogin) {
			p = getenv("TERM");
			cleanenv[0] = "PATH=:/usr/ucb:/bin:/usr/bin";
			cleanenv[1] = NULL;
			environ = cleanenv;
			(void)setenv("TERM", p, 1);
			if (chdir(pwd->pw_dir) < 0) {
				fprintf(stderr, "ksu: no directory\n");
				dest_tkt();
				exit(1);
			}
		}
		if (fulllogin || pwd->pw_uid)
			(void)setenv("USER", pwd->pw_name, 1);
		(void)setenv("HOME", pwd->pw_dir, 1);
		(void)setenv("SHELL", shell, 1);
#ifndef NO_GETUSERSHELL
	}
#endif

	if (iscsh == YES) {
		if (fastlogin)
			*np-- = "-f";
#ifndef NO_GETUSERSHELL
		if (asme)
			*np-- = "-m";
#endif
	}

	/* csh strips the first character... */
	*np = fulllogin ? "-ksu" : iscsh == YES ? "_ksu" : "ksu";

	if (pwd->pw_uid == 0)
		syslog(LOG_NOTICE|LOG_AUTH, "ksu: %s on %s",
		    username, mytty);
#ifndef NO_SETPRIORITY
	(void)setpriority(PRIO_PROCESS, 0, prio);
#endif
	execv(shell, np);
	fprintf(stderr, "ksu: no shell.\n");
	dest_tkt();
	exit(1);
}

#ifndef NO_GETUSERSHELL
chshell(sh)
	char *sh;
{
	char *cp, *getusershell();

	while ((cp = getusershell()) != NULL)
		if (!strcmp(cp, sh))
			return(1);
	return(0);
}
#endif /* NO_GETUSERSHELL */

koktologin(name, realm, user)
    char   *name;
    char   *realm;
    char   *user;
{
    struct auth_dat kdata_st;
    AUTH_DAT *kdata = &kdata_st;
    /* Cons together an AUTH_DAT structure for kuserok */
    memset((caddr_t) kdata, 0, sizeof(*kdata));
    strcpy(kdata->pname, name);
    strcpy(kdata->pinst, user);
    strcpy(kdata->prealm, realm);
    return (kuserok(kdata, user));
}
